File: Data Structures & Algorithms/generate-parentheses/submission-3.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        left, right = 0, 0
        res, sol = [], []

        def backtrack(left, right, sol):
            if left == right == n:
                res.append("".join(sol[:]))
                return

            if left < n:
                sol.append("(")
                backtrack(left + 1, right,  sol)
                sol.pop()

            if left > right:
                sol.append(")")
                backtrack(left, right + 1,  sol)
                sol.pop()
            
            return

        backtrack(left, right, sol)
        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-3.go
func IsPalindrome(s string) bool {
	// Convert to runes to correctly handle multi-byte Unicode characters
	runes := []rune(s)
	left := 0
	right := len(runes) - 1

	for left < right {
		if runes[left] != runes[right] {
			return false
		}
		left++
		right--
	}

	return true
}


func partition(s string) [][]string {
	sol := []string {}
	res := [][]string {}

	var backtrack func(i int) 

	backtrack = func(i int) {
        if i == len(s) {
            res = append(res, append([]string{}, sol...))
            return
        }

		for j := i; j < len(s); j++ {
            sub := s[i : j + 1]
			if IsPalindrome(sub) {
                sol = append(sol, sub)
                backtrack(j + 1)
                sol = sol[:len(sol) - 1]
            }
		}
	}

	backtrack(0)
	return res
}
